# Exploring eBay Car Sales Data

This project explores a dataset of used car listings from eBay Kleinanzeigen. The goal is to clean the dataset and perform an initial analysis to identify useful patterns in car prices, mileage, brands, and other characteristics.

In [2]:
import pandas as pd
import numpy as np

In [8]:
autos = pd.read_csv(r"E:\project_DS\Exploring eBay Car\autos.csv", encoding="latin1")
autos

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,kilometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-24 11:52:17,Golf_3_1.6,privat,Angebot,480,test,NaN,1993,manuell,0,golf,150000,0,benzin,volkswagen,NaN,2016-03-24 00:00:00,0,70435,2016-04-07 03:16:57
1,2016-03-24 10:58:45,A5_Sportback_2.7_Tdi,privat,Angebot,18300,test,coupe,2011,manuell,190,NaN,125000,5,diesel,audi,ja,2016-03-24 00:00:00,0,66954,2016-04-07 01:46:50
2,2016-03-14 12:52:21,"Jeep_Grand_Cherokee_""Overland""",privat,Angebot,9800,test,suv,2004,automatik,163,grand,125000,8,diesel,jeep,NaN,2016-03-14 00:00:00,0,90480,2016-04-05 12:47:46
3,2016-03-17 16:54:04,GOLF_4_1_4__3TÜRER,privat,Angebot,1500,test,kleinwagen,2001,manuell,75,golf,150000,6,benzin,volkswagen,nein,2016-03-17 00:00:00,0,91074,2016-03-17 17:40:17
4,2016-03-31 17:25:20,Skoda_Fabia_1.4_TDI_PD_Classic,privat,Angebot,3600,test,kleinwagen,2008,manuell,69,fabia,90000,7,diesel,skoda,nein,2016-03-31 00:00:00,0,60437,2016-04-06 10:17:21
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
371523,2016-03-14 17:48:27,Suche_t4___vito_ab_6_sitze,privat,Angebot,2200,test,NaN,2005,NaN,0,NaN,20000,1,NaN,sonstige_autos,NaN,2016-03-14 00:00:00,0,39576,2016-04-06 00:46:52
371524,2016-03-05 19:56:21,Smart_smart_leistungssteigerung_100ps,privat,Angebot,1199,test,cabrio,2000,automatik,101,fortwo,125000,3,benzin,smart,nein,2016-03-05 00:00:00,0,26135,2016-03-11 18:17:12
371525,2016-03-19 18:57:12,Volkswagen_Multivan_T4_TDI_7DC_UY2,privat,Angebot,9200,test,bus,1996,manuell,102,transporter,150000,3,diesel,volkswagen,nein,2016-03-19 00:00:00,0,87439,2016-04-07 07:15:26
371526,2016-03-20 19:41:08,VW_Golf_Kombi_1_9l_TDI,privat,Angebot,3400,test,kombi,2002,manuell,100,golf,150000,6,diesel,volkswagen,NaN,2016-03-20 00:00:00,0,40764,2016-03-24 12:45:21


In [7]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [9]:
autos.head()


,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,kilometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-24 11:52:17,Golf_3_1.6,privat,Angebot,480,test,NaN,1993,manuell,0,golf,150000,0,benzin,volkswagen,NaN,2016-03-24 00:00:00,0,70435,2016-04-07 03:16:57
1,2016-03-24 10:58:45,A5_Sportback_2.7_Tdi,privat,Angebot,18300,test,coupe,2011,manuell,190,NaN,125000,5,diesel,audi,ja,2016-03-24 00:00:00,0,66954,2016-04-07 01:46:50
2,2016-03-14 12:52:21,"Jeep_Grand_Cherokee_""Overland""",privat,Angebot,9800,test,suv,2004,automatik,163,grand,125000,8,diesel,jeep,NaN,2016-03-14 00:00:00,0,90480,2016-04-05 12:47:46
3,2016-03-17 16:54:04,GOLF_4_1_4__3TÜRER,privat,Angebot,1500,test,kleinwagen,2001,manuell,75,golf,150000,6,benzin,volkswagen,nein,2016-03-17 00:00:00,0,91074,2016-03-17 17:40:17
4,2016-03-31 17:25:20,Skoda_Fabia_1.4_TDI_PD_Classic,privat,Angebot,3600,test,kleinwagen,2008,manuell,69,fabia,90000,7,diesel,skoda,nein,2016-03-31 00:00:00,0,60437,2016-04-06 10:17:21


# Dataset Overview

*The dataset contains 371,528 rows and 20 columns.*

**Missing Values:**

Some columns contain missing values, including vehicleType, gearbox, model, fuelType, and notRepairedDamage.

**Data Types:**

The dataset contains both numerical and categorical data, and some categorical values are written in German.

In [10]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [ ]:
def clean_column_name(column):
    new_name = ""

    for char in column:
        if char.isupper():
            new_name += "_" + char.lower()
        else:
            new_name += char.lower()

    return new_name


# print(clean_column_name("dateCrawled"))
# print(clean_column_name("yearOfRegistration"))
# print(clean_column_name("notRepairedDamage"))

autos.columns = autos.columns.map(clean_column_name)
autos.rename(columns={"power_p_s": "power_ps"}, inplace=True)


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'kilometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [21]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'kilometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

# Step 3 — Initial Exploration and Cleaning

In [22]:
autos.describe(include="all")


,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,year_of_registration,gearbox,power_ps,model,kilometer,month_of_registration,fuel_type,brand,not_repaired_damage,date_created,nr_of_pictures,postal_code,last_seen
count,371528,371528,371528,371528,3.715280e+05,371528,333659,371528.000000,351319,371528.000000,351044,371528.000000,371528.000000,338142,371528,299468,371528,371528.0,371528.00000,371528
unique,280500,233531,2,2,NaN,2,8,NaN,2,NaN,251,NaN,NaN,7,40,2,114,NaN,NaN,182806
top,2016-03-24 14:49:47,Ford_Fiesta,privat,Angebot,NaN,test,limousine,NaN,manuell,NaN,golf,NaN,NaN,benzin,volkswagen,nein,2016-04-03 00:00:00,NaN,NaN,2016-04-07 06:45:59
freq,7,657,371525,371516,NaN,192585,95894,NaN,274214,NaN,30070,NaN,NaN,223857,79640,263182,14450,NaN,NaN,17
mean,NaN,NaN,NaN,NaN,1.729514e+04,NaN,NaN,2004.577997,NaN,115.549477,NaN,125618.688228,5.734445,NaN,NaN,NaN,NaN,0.0,50820.66764,NaN
std,NaN,NaN,NaN,NaN,3.587954e+06,NaN,NaN,92.866598,NaN,192.139578,NaN,40112.337051,3.712412,NaN,NaN,NaN,NaN,0.0,25799.08247,NaN
min,NaN,NaN,NaN,NaN,0.000000e+00,NaN,NaN,1000.000000,NaN,0.000000,NaN,5000.000000,0.000000,NaN,NaN,NaN,NaN,0.0,1067.00000,NaN
25%,NaN,NaN,NaN,NaN,1.150000e+03,NaN,NaN,1999.000000,NaN,70.000000,NaN,125000.000000,3.000000,NaN,NaN,NaN,NaN,0.0,30459.00000,NaN
50%,NaN,NaN,NaN,NaN,2.950000e+03,NaN,NaN,2003.000000,NaN,105.000000,NaN,150000.000000,6.000000,NaN,NaN,NaN,NaN,0.0,49610.00000,NaN
75%,NaN,NaN,NaN,NaN,7.200000e+03,NaN,NaN,2008.000000,NaN,150.000000,NaN,150000.000000,9.000000,NaN,NaN,NaN,NaN,0.0,71546.00000,NaN


In [23]:
autos["seller"].unique()


array(['privat', 'gewerblich'], dtype=object)

In [24]:
autos["offer_type"].unique()


array(['Angebot', 'Gesuch'], dtype=object)

In [27]:
print(autos["seller"].value_counts())
print("##############################")
print(autos["offer_type"].value_counts())


seller
privat        371525
gewerblich         3
Name: count, dtype: int64
##############################
offer_type
Angebot    371516
Gesuch         12
Name: count, dtype: int64


In [28]:
autos.isnull().sum()


date_crawled                 0
name                         0
seller                       0
offer_type                   0
price                        0
abtest                       0
vehicle_type             37869
year_of_registration         0
gearbox                  20209
power_ps                     0
model                    20484
kilometer                    0
month_of_registration        0
fuel_type                33386
brand                        0
not_repaired_damage      72060
date_created                 0
nr_of_pictures               0
postal_code                  0
last_seen                    0
dtype: int64

In [29]:
autos.duplicated().sum()


4

In [30]:
autos[autos.duplicated()]


,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,year_of_registration,gearbox,power_ps,model,kilometer,month_of_registration,fuel_type,brand,not_repaired_damage,date_created,nr_of_pictures,postal_code,last_seen
179448,2016-03-08 18:42:48,Mercedes_Benz_CLK_Coupe_230_Kompressor_Sport,privat,Angebot,1799,test,coupe,1999,automatik,193,clk,20000,7,benzin,mercedes_benz,nein,2016-03-08 00:00:00,0,89518,2016-03-09 09:46:57
242556,2016-03-28 00:56:10,Suzuki_Ignis,privat,Angebot,1000,control,kleinwagen,2002,manuell,83,andere,150000,1,benzin,suzuki,nein,2016-03-28 00:00:00,0,66589,2016-03-28 08:46:21
270719,2016-04-03 09:01:15,Mercedes_Benz_CLK_320_W209,privat,Angebot,4699,test,coupe,2003,automatik,218,clk,125000,6,benzin,mercedes_benz,ja,2016-04-03 00:00:00,0,75196,2016-04-07 09:44:54
341474,2016-03-18 18:46:15,Volkswagen_Passat_Variant_1.9_TDI_Highline,privat,Angebot,1999,control,kombi,2001,manuell,131,passat,150000,7,diesel,volkswagen,nein,2016-03-18 00:00:00,0,36391,2016-03-18 18:46:15


In [31]:
autos[autos.duplicated(keep=False)]


,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,year_of_registration,gearbox,power_ps,model,kilometer,month_of_registration,fuel_type,brand,not_repaired_damage,date_created,nr_of_pictures,postal_code,last_seen
43501,2016-03-18 18:46:15,Volkswagen_Passat_Variant_1.9_TDI_Highline,privat,Angebot,1999,control,kombi,2001,manuell,131,passat,150000,7,diesel,volkswagen,nein,2016-03-18 00:00:00,0,36391,2016-03-18 18:46:15
92386,2016-03-08 18:42:48,Mercedes_Benz_CLK_Coupe_230_Kompressor_Sport,privat,Angebot,1799,test,coupe,1999,automatik,193,clk,20000,7,benzin,mercedes_benz,nein,2016-03-08 00:00:00,0,89518,2016-03-09 09:46:57
95425,2016-03-28 00:56:10,Suzuki_Ignis,privat,Angebot,1000,control,kleinwagen,2002,manuell,83,andere,150000,1,benzin,suzuki,nein,2016-03-28 00:00:00,0,66589,2016-03-28 08:46:21
179448,2016-03-08 18:42:48,Mercedes_Benz_CLK_Coupe_230_Kompressor_Sport,privat,Angebot,1799,test,coupe,1999,automatik,193,clk,20000,7,benzin,mercedes_benz,nein,2016-03-08 00:00:00,0,89518,2016-03-09 09:46:57
196894,2016-04-03 09:01:15,Mercedes_Benz_CLK_320_W209,privat,Angebot,4699,test,coupe,2003,automatik,218,clk,125000,6,benzin,mercedes_benz,ja,2016-04-03 00:00:00,0,75196,2016-04-07 09:44:54
242556,2016-03-28 00:56:10,Suzuki_Ignis,privat,Angebot,1000,control,kleinwagen,2002,manuell,83,andere,150000,1,benzin,suzuki,nein,2016-03-28 00:00:00,0,66589,2016-03-28 08:46:21
270719,2016-04-03 09:01:15,Mercedes_Benz_CLK_320_W209,privat,Angebot,4699,test,coupe,2003,automatik,218,clk,125000,6,benzin,mercedes_benz,ja,2016-04-03 00:00:00,0,75196,2016-04-07 09:44:54
341474,2016-03-18 18:46:15,Volkswagen_Passat_Variant_1.9_TDI_Highline,privat,Angebot,1999,control,kombi,2001,manuell,131,passat,150000,7,diesel,volkswagen,nein,2016-03-18 00:00:00,0,36391,2016-03-18 18:46:15


In [32]:
autos.drop_duplicates(inplace=True)
autos.duplicated().sum()


0

In [33]:
autos.shape


(371524, 20)

# Step 4 — Exploring the Odometer and Price Columns

In [35]:
autos["kilometer"].value_counts().sort_index()


kilometer
5000        7069
10000       1949
20000       5675
30000       6041
40000       6376
50000       7615
60000       8669
70000       9773
80000      11053
90000      12523
100000     15920
125000     38066
150000    240795
Name: count, dtype: int64

In [36]:
autos["price"].value_counts().sort_index()


price
0             10778
1              1189
2                12
3                 8
4                 1
              ...  
32545461          1
74185296          1
99000000          1
99999999         15
2147483647        1
Name: count, Length: 5597, dtype: int64

In [37]:
autos = autos[autos["price"].between(1, 350000)]
autos.shape


(360631, 20)

In [38]:
autos["price"].describe()


count    360631.000000
mean       5898.711048
std        8866.399857
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Step 5: Exploring the Date Columns

In [39]:
autos[["date_crawled", "date_created", "last_seen"]].head()


,date_crawled,date_created,last_seen
0,2016-03-24 11:52:17,2016-03-24 00:00:00,2016-04-07 03:16:57
1,2016-03-24 10:58:45,2016-03-24 00:00:00,2016-04-07 01:46:50
2,2016-03-14 12:52:21,2016-03-14 00:00:00,2016-04-05 12:47:46
3,2016-03-17 16:54:04,2016-03-17 00:00:00,2016-03-17 17:40:17
4,2016-03-31 17:25:20,2016-03-31 00:00:00,2016-04-06 10:17:21


In [40]:
autos[["date_crawled", "date_created", "last_seen"]].dtypes


date_crawled    object
date_created    object
last_seen       object
dtype: object

In [41]:
autos["date_crawled"] = pd.to_datetime(autos["date_crawled"])
autos["date_created"] = pd.to_datetime(autos["date_created"])
autos["last_seen"] = pd.to_datetime(autos["last_seen"])

autos[["date_crawled", "date_created", "last_seen"]].dtypes


C:\Users\anasm\AppData\Local\Temp\ipykernel_4208\4260932238.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["date_crawled"] = pd.to_datetime(autos["date_crawled"])
C:\Users\anasm\AppData\Local\Temp\ipykernel_4208\4260932238.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["date_created"] = pd.to_datetime(autos["date_created"])
C:\Users\anasm\AppData\Local\Temp\ipykernel_4208\4260932238.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try us

date_crawled    datetime64[ns]
date_created    datetime64[ns]
last_seen       datetime64[ns]
dtype: object

In [43]:
print(autos["date_crawled"].min())
print(autos["date_crawled"].max())

print(autos["date_created"].min())
print(autos["date_created"].max())

print(autos["last_seen"].min())
print(autos["last_seen"].max())


2016-03-05 14:06:22
2016-04-07 14:36:58
2014-03-10 00:00:00
2016-04-07 00:00:00
2016-03-05 14:15:08
2016-04-07 14:58:51


In [44]:
autos["date_created"].dt.year.value_counts().sort_index()


date_created
2014         1
2015        29
2016    360601
Name: count, dtype: int64

In [45]:
autos["date_created"].dt.month.value_counts().sort_index()


date_created
1         75
2        415
3     301696
4      58417
6          2
8          2
9          3
10         1
11        10
12        10
Name: count, dtype: int64

# Step 6 — Incorrect Registration Year Data 

In [46]:
autos["year_of_registration"].value_counts().sort_index()


year_of_registration
1000    32
1001     1
1039     1
1111     1
1234     4
        ..
8500     1
8888     2
9000     4
9450     1
9999    18
Name: count, Length: 145, dtype: int64

In [47]:
invalid_years = ~autos["year_of_registration"].between(1900, 2016)

print(invalid_years.sum())


13975


In [48]:
autos = autos[autos["year_of_registration"].between(1900, 2016)].copy()
autos["year_of_registration"].describe()


count    346656.000000
mean       2002.896508
std           7.244757
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: year_of_registration, dtype: float64

In [49]:
autos.shape


(346656, 20)

# Step 7 — Exploring Price by Brand 

In [50]:
autos.groupby("brand")["price"].mean()


brand
alfa_romeo         4291.759945
audi               9086.279598
bmw                8449.121823
chevrolet          7117.870190
chrysler           4121.688574
citroen            3734.280604
dacia              5922.862428
daewoo             1027.841584
daihatsu           1775.097463
fiat               2892.953514
ford               3696.405969
honda              4005.660283
hyundai            5567.050043
jaguar            13765.309484
jeep              11213.438158
kia                5855.187735
lada               3191.425121
lancia             3289.995565
land_rover        17070.102530
mazda              4076.357960
mercedes_benz      8551.966040
mini              10080.620658
mitsubishi         3407.478809
nissan             4708.931269
opel               2971.899913
peugeot            3267.612647
porsche           42258.824431
renault            2437.605452
rover              1600.442177
saab               3955.594488
seat               4541.999227
skoda              6530.510850
sm

In [51]:
brand_mean_price = autos.groupby("brand")["price"].mean().sort_values(ascending=False)

brand_mean_price


brand
porsche           42258.824431
land_rover        17070.102530
sonstige_autos    14288.713461
jaguar            13765.309484
jeep              11213.438158
mini              10080.620658
audi               9086.279598
mercedes_benz      8551.966040
bmw                8449.121823
chevrolet          7117.870190
skoda              6530.510850
dacia              5922.862428
kia                5855.187735
hyundai            5567.050043
volkswagen         5400.235110
toyota             5339.680571
volvo              5238.420372
nissan             4708.931269
seat               4541.999227
subaru             4386.149864
alfa_romeo         4291.759945
chrysler           4121.688574
mazda              4076.357960
suzuki             4045.770975
honda              4005.660283
saab               3955.594488
citroen            3734.280604
ford               3696.405969
smart              3632.480878
mitsubishi         3407.478809
lancia             3289.995565
peugeot            3267.612647
la

# Step 8 — Storing Aggregate Data in a DataFrame

In [53]:
autos.groupby("brand")[["price"]].mean()


,price
brand,
alfa_romeo,4291.759945
audi,9086.279598
bmw,8449.121823
chevrolet,7117.870190
chrysler,4121.688574
citroen,3734.280604
dacia,5922.862428
daewoo,1027.841584
daihatsu,1775.097463


In [52]:
brand_mean_prices = autos.groupby("brand")[["price"]].mean()

print(type(brand_mean_prices))
brand_mean_prices.head()


<class 'pandas.core.frame.DataFrame'>


,price
brand,
alfa_romeo,4291.759945
audi,9086.279598
bmw,8449.121823
chevrolet,7117.870190
chrysler,4121.688574
